src/chainstate/stacks/index/node.rs

```diff
@@ -0,0 +1,2756 @@
+};
+
+use chainstate::stacks::index::{
+    TrieHash,
+    TRIEHASH_ENCODED_SIZE,
+    fast_extend_from_slice,
+    slice_partialeq,
+    MARFValue, 
+    MARF_VALUE_ENCODED_SIZE,
+};
+
+use chainstate::stacks::index::Error as Error;
+
+use util::hash::to_hex;
+use util::log;
+
+/// All numeric values of a Trie node when encoded.
+/// They are all 7-bit numbers -- the 8th bit is used to indicate whether or not the value
+/// identifies a back-pointer to be followed.
+pub mod TrieNodeID {
+    pub const Empty : u8 = 0;
+    pub const Leaf: u8 = 1;
+    pub const Node4 : u8 = 2;
+    pub const Node16 : u8 = 3;
+    pub const Node48 : u8 = 4;
+    pub const Node256 : u8 = 5;
+}
+
+/// A node ID encodes a back-pointer if its high bit is set
+pub fn is_backptr(id: u8) -> bool {
+    id & 0x80 != 0
+}
+
+/// Set the back-pointer bit
+pub fn set_backptr(id: u8) -> u8 {
+    id | 0x80
+}
+
+/// Clear the back-pointer bit
+pub fn clear_backptr(id: u8) -> u8 {
+    id & 0x7f
+}
+
+/// A path in the Trie is the SHA2-512/256 hash of its key.
+pub struct TriePath([u8; 32]);
+impl_array_newtype!(TriePath, u8, 32);
+impl_array_hexstring_fmt!(TriePath);
+impl_byte_array_newtype!(TriePath, u8, 32);
+
+pub const TRIEPATH_MAX_LEN : usize = 32;
+
+impl TriePath {
+    pub fn from_key(k: &String) -> TriePath {
+        let h = TrieHash::from_data(k.as_bytes());
+        let mut hb = [0u8; TRIEPATH_MAX_LEN];
+        hb.copy_from_slice(h.as_bytes());
+        TriePath(hb)
+    }
+}
+
+/// All Trie nodes implement the following methods:
+pub trait TrieNode {
+    /// Node ID for encoding/decoding
+    fn id(&self) -> u8;
+
+    /// Is the node devoid of children?
+    fn empty() -> Self;
+
+    /// Follow a path character to a child pointer
+    fn walk(&self, chr: u8) -> Option<TriePtr>;
+
+    /// Insert a child pointer if the path character slot is not occupied.
+    /// Return true if inserted, false if the slot is already filled
+    fn insert(&mut self, ptr: &TriePtr) -> bool;
+
+    /// Replace an existing child pointer with a new one.  Returns true if replaced; false if the
+    /// child does not exist.
+    fn replace(&mut self, ptr: &TriePtr) -> bool;
+
+    /// Read an encoded instance of this node from a byte stream and instantiate it.
+    fn from_bytes<R: Read>(r: &mut R) -> Result<Self, Error>
+        where Self: std::marker::Sized;
+
+    /// Encode this node instance into a byte stream and write it to the given memory buffer.
+    fn to_bytes(&self, buf: &mut Vec<u8>) -> ();
+
+    /// Encode the consensus-relevant bytes of this node and write it to the given memory buffer.
+    fn to_consensus_bytes(&self, &mut Vec<u8>) -> ();
+
+    /// Calculate how many bytes this node will take to encode.
+    fn byte_len(&self) -> usize;
+
+    /// Get a reference to the children of this node.
+    fn ptrs(&self) -> &[TriePtr];
+
+    // this is a way to construct a TrieNodeType from an object that implements this trait
+    // DO NOT USE DIRECTLY
```

Comment:
I think adding `as_trie_node_type()` would eliminate a lot of duplication in trie.rs and node.r (I don't think all these try_as's would be necessary).

Reply:
I've managed to "contain" this within an impl for the TrieNode trait.  Let me know what you think?